aws/sdk/integration-tests/s3/tests/throughput-timeout.rs

```diff
@@ -0,0 +1,142 @@
+
+#[should_panic = "minimum throughput was specified at 2 B/s, but throughput of 1.5 B/s was observed"]
+#[tokio::test]
+async fn test_throughput_timeout_less_than() {
+    let _logs = capture_test_logs();
+    let (time_source, sleep) = instant_time_and_sleep(UNIX_EPOCH);
+
+    let shared_sleep = sleep.clone();
+    // Will send ~1 byte per second.
+    let stream = futures_util::stream::unfold(1, move |state| {
+        let sleep = shared_sleep.clone();
+        async move {
+            if state > 255 {
+                None
+            } else {
+                sleep.sleep(Duration::from_secs(1)).await;
+                Some((
+                    Result::<_, Infallible>::Ok(Bytes::from(vec![state as u8])),
+                    state + 1,
+                ))
+            }
+        }
+    });
+    let body = ByteStream::new(SdkBody::from(hyper::body::Body::wrap_stream(stream)));
+    let body = body.map(move |body| {
+        let ts = time_source.clone();
+        // Throw an error if the stream sends less than 2 bytes per second
+        let minimum_throughput = (2u64, Duration::from_secs(1));
+        SdkBody::from_dyn(aws_smithy_http::body::BoxBody::new(
+            MinimumThroughputBody::new(ts, body, minimum_throughput),
+        ))
+    });
+    let res = body.collect().await;
+
+    match res {
+        Ok(_) => panic!("Expected an error due to slow stream but no error occurred."),
+        Err(e) => panic!("{}", DisplayErrorContext(e)),
+    }
+}
+
+const EXPECTED_BYTES: Lazy<Vec<u8>> = Lazy::new(|| (1..=255).map(|i| i as u8).collect::<Vec<_>>());
+
+#[tokio::test]
+async fn test_throughput_timeout_equal_to() {
+    let _logs = capture_test_logs();
+    let (time_source, sleep) = instant_time_and_sleep(UNIX_EPOCH);
+    let time_source: SharedTimeSource = time_source.into_shared();
+    let sleep: SharedAsyncSleep = sleep.into_shared();
+
+    // Will send ~1 byte per second.
+    let stream = futures_util::stream::unfold(1, move |state| {
+        let sleep = sleep.clone();
+        async move {
+            if state > 255 {
+                None
+            } else {
+                sleep.sleep(Duration::from_secs(1)).await;
+                Some((
+                    Result::<_, Infallible>::Ok(Bytes::from(vec![state as u8])),
+                    state + 1,
+                ))
+            }
+        }
+    });
+    let body = ByteStream::new(SdkBody::from(hyper::body::Body::wrap_stream(stream)));
+    let body = body.map(move |body| {
+        let time_source = time_source.clone();
+        // Throw an error if the stream sends less than 1 byte per second
+        let minimum_throughput = (1u64, Duration::from_secs(1));
+        SdkBody::from_dyn(aws_smithy_http::body::BoxBody::new(
+            MinimumThroughputBody::new(time_source, body, minimum_throughput),
+        ))
+    });
+    // assert_eq!(255.0, time_source.seconds_since_unix_epoch());
+    // assert_eq!(Duration::from_secs(255), sleep.total_duration());
```

Comment:
I haven't yet figure out how to retain access to the `sleep_impl` and `time_source` such that I can run these asserts.

Reply:
Should be able to clone them before calling `into_shared()`.